File: companion/ios/CanRosettaCompanion/Sensors/LocationSource.swift
import Foundation
import CoreLocation
import os

/// One line of `phone/location.jsonl`. Field names/units follow
/// `docs/data-format.md` and `schemas/location.record.schema.json`:
///
/// - `lat`, `lon`   degrees                                — required
/// - `alt`          altitude, m
/// - `speed`        m/s over ground, `-1` if unknown
/// - `course`       degrees from true north, `-1` if unknown
/// - `hAcc`/`vAcc`  horizontal/vertical accuracy, m
/// - `producedByAccessory`  optional; true when the fix came from an external
///   accessory rather than the phone's own GNSS (see below)
///
/// `tUtc` encodes to `t_utc`, `hAcc`/`vAcc` to `h_acc`/`v_acc`,
/// `producedByAccessory` to `produced_by_accessory` (omitted when nil — the
/// synthesized `Encodable` uses `encodeIfPresent` for optionals).
struct LocationRecord: Encodable {
    let tUtc: Double
    let lat: Double
    let lon: Double
    let alt: Double
    let speed: Double
    let course: Double
    let hAcc: Double
    let vAcc: Double
    /// Wireless CarPlay head units transparently feed vehicle GNSS (typically
    /// wheel-speed-aided dead reckoning) into CoreLocation, so a docked
    /// iPhone's "GPS" may actually be the car's — not an independent phone
    /// reference. The pipeline needs this flag to tell fused fixes from
    /// phone-only GPS. Written only when CoreLocation reports source
    /// information (iOS 15+ API; our floor is 16, but the field itself can be
    /// absent), so `nil` here means "provenance unknown", not "phone GPS".
    let producedByAccessory: Bool?
}

/// Wraps `CLLocationManager` at best accuracy for driving, mapping each fix to
/// a `LocationRecord`. Handles authorization. Emits raw fixes only — no
/// smoothing or interpolation.
final class LocationSource: NSObject, CLLocationManagerDelegate {

    private let manager = CLLocationManager()
    private let logger = Logger(subsystem: AppInfo.subsystem, category: "location")

    /// Called for every fix, on the main queue (CoreLocation's delegate queue).
    var onRecord: ((LocationRecord) -> Void)?
    /// Called when authorization status changes.
    var onAuthorizationChange: ((CLAuthorizationStatus) -> Void)?

    /// Latest reported horizontal accuracy (m), or `nil` if no valid fix yet.
    /// Read from the main thread for the live UI.
    private(set) var lastHorizontalAccuracy: Double?

    /// Whether we want background updates (requires "Always" auth + the
    /// `location` background mode). Recording while the screen is off / app
    /// backgrounded is common in a cradle, so default to true.
    var wantsBackgroundUpdates = true

    override init() {
        super.init()
        manager.delegate = self
        manager.desiredAccuracy = kCLLocationAccuracyBestForNavigation
        manager.distanceFilter = kCLDistanceFilterNone
        manager.activityType = .automotiveNavigation
        manager.pausesLocationUpdatesAutomatically = false
    }

    var authorizationStatus: CLAuthorizationStatus { manager.authorizationStatus }

    /// Ask for the strongest authorization we can. Call before `start()`.
    func requestAuthorization() {
        switch manager.authorizationStatus {
        case .notDetermined:
            // "When in use" first; iOS will offer the upgrade to "Always".
            manager.requestWhenInUseAuthorization()
        case .authorizedWhenInUse:
            if wantsBackgroundUpdates { manager.requestAlwaysAuthorization() }
        default:
            break
        }
    }

    func start() {
        let status = manager.authorizationStatus
        guard status == .authorizedWhenInUse || status == .authorizedAlways else {
            logger.error("Cannot start location: not authorized (status \(status.rawValue))")
            return
        }
        // Background updates are only legal with "Always" + the background mode.
        if wantsBackgroundUpdates, status == .authorizedAlways {
            manager.allowsBackgroundLocationUpdates = true
            manager.showsBackgroundLocationIndicator = true
        }
        manager.startUpdatingLocation()
        logger.info("Started location updates")
    }

    func stop() {
        manager.stopUpdatingLocation()
        manager.allowsBackgroundLocationUpdates = false
        logger.info("Stopped location updates")
    }

    // MARK: - CLLocationManagerDelegate

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        logger.info("Location authorization changed: \(manager.authorizationStatus.rawValue)")
        onAuthorizationChange?(manager.authorizationStatus)
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        for loc in locations {
            // Use the fix's own timestamp: CoreLocation stamps each fix at
            // acquisition time on the same wall clock, and it is the most
            // accurate time we have for a GPS sample.
            let record = LocationRecord(
                tUtc: loc.timestamp.timeIntervalSince1970,
                lat: loc.coordinate.latitude,
                lon: loc.coordinate.longitude,
                alt: loc.altitude,
                speed: loc.speed >= 0 ? loc.speed : -1,
                course: loc.course >= 0 ? loc.course : -1,
                hAcc: loc.horizontalAccuracy,
                vAcc: loc.verticalAccuracy,
                // GPS provenance: `sourceInformation` is nil when CoreLocation
                // has nothing to say (e.g. Simulator); only then do we omit the
                // field, keeping "unknown" distinct from "phone GPS" (false).
                producedByAccessory: loc.sourceInformation?.isProducedByAccessory
            )
            if loc.horizontalAccuracy >= 0 {
                lastHorizontalAccuracy = loc.horizontalAccuracy
            }
            onRecord?(record)
        }
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        logger.error("Location error: \(error.localizedDescription)")
    }
}

File: companion/android/app/src/main/java/com/inomotech/canrosetta/companion/sensors/LocationSource.kt
package com.inomotech.canrosetta.companion.sensors

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.os.Looper
import android.util.Log
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.inomotech.canrosetta.companion.AppInfo
import com.inomotech.canrosetta.companion.io.Records
import com.inomotech.canrosetta.companion.time.TimeMath
import org.json.JSONObject

/**
 * Wraps the fused location provider at high accuracy for driving, mapping each
 * fix to a `phone/location.jsonl` record. Emits raw fixes only — no smoothing.
 *
 * Units follow the schema: `speed` m/s (`-1` if unknown), `course` deg from true
 * north (`-1` if unknown), `h_acc`/`v_acc` meters. `t_utc` comes from the fix's
 * own wall-clock time (`Location.time`), the best time available for a GPS sample.
 */
class LocationSource(context: Context) {

    private val client: FusedLocationProviderClient =
        LocationServices.getFusedLocationProviderClient(context)

    /** Called for every fix, on the main looper. */
    var onRecord: ((JSONObject) -> Unit)? = null

    /** Latest reported horizontal accuracy (m), or null if no valid fix yet. */
    @Volatile
    var lastHorizontalAccuracy: Double? = null
        private set

    private val callback = object : LocationCallback() {
        override fun onLocationResult(result: LocationResult) {
            for (loc in result.locations) emit(loc)
        }
    }

    private fun emit(loc: Location) {
        val speed = if (loc.hasSpeed()) loc.speed.toDouble() else -1.0
        val course = if (loc.hasBearing()) loc.bearing.toDouble() else -1.0
        val hAcc = if (loc.hasAccuracy()) loc.accuracy.toDouble() else -1.0
        val vAcc = if (loc.hasVerticalAccuracy()) loc.verticalAccuracyMeters.toDouble() else -1.0
        if (hAcc >= 0) lastHorizontalAccuracy = hAcc
        val record = Records.location(
            tUtc = TimeMath.millisToUtc(loc.time),
            lat = loc.latitude,
            lon = loc.longitude,
            alt = loc.altitude,
            speed = speed,
            course = course,
            hAcc = hAcc,
            vAcc = vAcc,
        )
        onRecord?.invoke(record)
    }

    @SuppressLint("MissingPermission")
    fun start() {
        val request = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 1000L)
            .setMinUpdateIntervalMillis(100L)
            .build()
        try {
            client.requestLocationUpdates(request, callback, Looper.getMainLooper())
        } catch (e: SecurityException) {
            Log.e(AppInfo.TAG, "Cannot start location: permission missing (${e.message})")
        }
    }

    fun stop() {
        client.removeLocationUpdates(callback)
    }
}
